File: windows_phone/MpOS API/MpOS/Net/Profile/ProfileNetworkFull.cs
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
using System;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using Ufc.MpOS.Net.Endpoint;
using Ufc.MpOS.Net.Profile.Model;
using Ufc.MpOS.Util;
using Ufc.MpOS.Util.Exceptions;
using Windows.Networking.Sockets;
using Windows.Storage.Streams;

namespace Ufc.MpOS.Net.Profile
{
	sealed class ProfileNetworkFull : ProfileNetworkTask
	{
		private byte[] data;//used on uploading...

		public ProfileNetworkFull(ServerContent server) : base(server, "ProfileFull Started on endpoint: " + server.Ip) {
			data = new byte[32 * 1024];
			new Random().NextBytes(data);
		}

		/**
		 * Feedback code:
		 * 15 -> Finished Ping TCP Test
		 * 30 -> Finished Ping UDP Test
		 * 35 -> Finished Ping Test with packet loss
		 * 50 -> Finished Jitter Calculation
		 * 55 -> Start Donwload Test
		 * 75 -> Start Upload Test
		 * 100 -> Finished Conection Test
		 */
		protected override Network DoInBackground()
		{
			System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: " + startMessage);
			PublishProgress(0);

			try
			{
				var pingTask = new Task<long[]>[] { PingService(true) };
				Task<long[]>.WaitAll(pingTask, 15000);
				PublishProgress(15);
				network.ResultPingTcp = pingTask[0].Result;

				pingTask = new Task<long[]>[] { PingService(false) };
				Task<long[]>.WaitAll(pingTask, 15000);
				PublishProgress(30);
				network.ResultPingUdp = pingTask[0].Result;

				pingTask = null;
				if (halted)
				{
					return null;
				}

				network.LossPacket = LossPacketCalculation(network);
				PublishProgress(35);

				var jitterTask = new Task []{ JitterCalculation() };
				Task.WaitAll(jitterTask, 15000);
				if (halted)
				{
					return null;
				}
				jitterTask = new Task[] { RetrieveJitterResult() };
				Task.WaitAll(jitterTask, 15000);
				
				jitterTask = null;
				PublishProgress(50);

				var bandwidthTask = new Task<bool>[] { BandwidthCalculation() };
				Task<bool>.WaitAll(bandwidthTask, 120000);
				bool finished = bandwidthTask[0].Result;
				if (halted || !finished)
				{
					return null;
				}
				bandwidthTask = null;

				var sent = SentNetworkProfile(network);

				PublishProgress(100);
				return network;
			}
			catch (AggregateException e)
			{
				e.Handle((ex) =>
				{
					if (ex is IOException)
					{
						System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: I/O Error -> " + ex.ToString());
						return true;
					}
					else if (ex is InterruptedException)
					{
						System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: InterruptedException -> " + ex.Message);
						return true;
					}
					else if (ex is Exception)
					{
						System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: Exception -> " + ex.Message);
						return true;
					}

					return false;
				});
			}
			finally
			{
				System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: Finished Profile Task");
			}

			PublishProgress(100);
			return null;
		}

		private async Task JitterCalculation()
		{
			datagramSocket = new DatagramSocket();

			datagramSocket.MessageReceived += delegate(DatagramSocket socket, DatagramSocketMessageReceivedEventArgs args)
			{
				System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: Jitter Finish");
			};

			string port = server.JitterTestPort.ToString();
			await datagramSocket.BindServiceNameAsync(port);
			using (DataWriter writer = new DataWriter(await datagramSocket.GetOutputStreamAsync(hostName, port)))
			{
				for (int i = 0; i < 21; i++)
				{
					writer.WriteString("jitter");
					await writer.StoreAsync();
					await writer.FlushAsync();

					await Task.Delay(250);
				}
			}

			Close(ref datagramSocket);
		}

		private async Task RetrieveJitterResult()
		{	
			streamSocket = new StreamSocket();
			await streamSocket.ConnectAsync(hostName, server.JitterRetrieveResultPort.ToString());

			await Task.Delay(1500);

			DataWriter writer = new DataWriter(streamSocket.OutputStream);
			DataReader reader = new DataReader(streamSocket.InputStream);

			writer.WriteString("get");
			await writer.StoreAsync();
			await writer.FlushAsync();

			reader.InputStreamOptions = InputStreamOptions.Partial;
			var read = await reader.LoadAsync(16);

			string message = reader.ReadString(read);
			network.Jitter = Convert.ToInt32(message);

			Close(ref writer);
			Close(ref reader);
			Close(ref streamSocket);
		}

		private async Task<bool> BandwidthCalculation()
		{
			System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: Download");
			PublishProgress(55);

			streamSocket = new StreamSocket();
			await streamSocket.ConnectAsync(hostName, server.BandwidthPort.ToString());

			DataWriter writer = new DataWriter(streamSocket.OutputStream);
			DataReader reader = new DataReader(streamSocket.InputStream);

			byte[] endDown = Encoding.UTF8.GetBytes("end_down");
			byte[] endSession = Encoding.UTF8.GetBytes("end_session");

			writer.WriteString("down");
			await writer.StoreAsync();
			await writer.FlushAsync();

			uint bufferSize = 8192;
			reader.InputStreamOptions = InputStreamOptions.Partial;

			long countBytes = 0L;
			byte[] downloadData = new byte[bufferSize];
			while (true)
			{
				uint readTransmittedData = await reader.LoadAsync(bufferSize);
				countBytes += readTransmittedData;

				if (readTransmittedData == bufferSize)
				{
					reader.ReadBytes(downloadData);
					if (GetDonwloadBandwidth(countBytes, ref downloadData, endDown))
					{
						break;
					}
				}
				else
				{
					byte[] tempData = new byte[readTransmittedData];
					reader.ReadBytes(tempData);
					//System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: " + readTransmittedData + "/" + countBytes);
					if (GetDonwloadBandwidth(countBytes, ref tempData, endDown))
					{
						break;
					}
				}
			}

			if (halted)
			{
				return false;
			}

			System.Diagnostics.Debug.WriteLine("[ProfileNetworkFull]: Upload");
			PublishProgress(75);
			writer.WriteString("up");
			await writer.StoreAsync();
			await writer.FlushAsync();

			System.Diagnostics.Stopwatch stopWatch = System.Diagnostics.Stopwatch.StartNew();
			while (stopWatch.ElapsedMilliseconds < 11000)
			{
				writer.WriteBytes(data);
				await writer.StoreAsync();
				await writer.FlushAsync();
			}
			stopWatch.Stop();

			writer.WriteString("end_up");
			await writer.StoreAsync();
			await writer.FlushAsync();

			var readFinish = await reader.LoadAsync(64);
			var endSessionBuffer = new byte[readFinish];
			reader.ReadBytes(endSessionBuffer);

			if (FunctionUtil.ContainsArrays(endSessionBuffer, endSession))
			{
				string dataBlock = Encoding.UTF8.GetString(endSessionBuffer, 0, endSessionBuffer.Length);
				string[] res = dataBlock.Split(new char[]{':'});
				network.BandwidthUpload = res[1];
			}

			Close(ref writer);
			Close(ref reader);
			Close(ref streamSocket);

			endSessionBuffer = null;

			return true;
		}

		private bool GetDonwloadBandwidth(long countBytes, ref byte[] downloadData, byte[] endDown)
		{
			if (FunctionUtil.ContainsArrays(downloadData, endDown))
			{
				// bytes * 8bits / 7s * 1E+6 = X Mbits
				double bandwidth = ((double)(countBytes * 8L) / (double)(7.0 * 1E+6));
				network.BandwidthDownload = Convert.ToString(bandwidth);
				downloadData = null;

				return true;
			}

			return false;
		}
	}
}

File: android/MpOS API/src/br/ufc/mdcc/mpos/net/profile/ProfileNetworkFull.java
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package br.ufc.mdcc.mpos.net.profile;

import java.io.IOException;
import java.util.Random;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.Semaphore;

import android.util.Log;
import br.ufc.mdcc.mpos.net.Protocol;
import br.ufc.mdcc.mpos.net.core.ClientAbstract;
import br.ufc.mdcc.mpos.net.core.FactoryClient;
import br.ufc.mdcc.mpos.net.core.ReceiveDataEvent;
import br.ufc.mdcc.mpos.net.endpoint.ServerContent;
import br.ufc.mdcc.mpos.net.exceptions.MissedEventException;
import br.ufc.mdcc.mpos.net.profile.model.Network;
import br.ufc.mdcc.mpos.util.TaskResult;
import br.ufc.mdcc.mpos.util.Util;

/**
 * This implement a full profile client.
 * 
 * @author Philipp B. Costa
 */
public final class ProfileNetworkFull extends ProfileNetworkTask {
	private byte data[] = new byte[32 * 1024];

	private Network network;
	private boolean bandwidthDone = false;

	public ProfileNetworkFull(TaskResult<Network> result, ServerContent server) throws MissedEventException {
		super(server, result, ProfileNetworkFull.class, "ProfileFull Started on endpoint: " + server.getIp());

		// randomize os dados que serão enviados
		new Random().nextBytes(data);
	}
	/**
     * Feedback code:
     * 15 -> Finished Ping TCP Test
     * 30 -> Finished Ping UDP Test
     * 35 -> Finished Ping Test with packet loss
     * 50 -> Finished Jitter Calculation
     * 55 -> Start Donwload Test
     * 75 -> Start Upload Test
     * 100 -> Finished Conection Test
     */
	@Override
	protected Network doInBackground(Void... params) {
		network = new Network();

		try {
			Log.i(clsName, "ping tcp");
			long[] pings = pingService(Protocol.TCP_EVENT);
			network.setResultPingTcp(pings);
			publishProgress(15);

			Log.i(clsName, "ping udp");
			pings = pingService(Protocol.UDP_EVENT);
			network.setResultPingUdp(pings);
			publishProgress(30);

			Log.i(clsName, "loss packet udp");
			// conta os pacotes perdidos UDP

			if (halted) {
				return null;
			}

			network.setLossPacket(lossPacketCalculation(network));
			publishProgress(35);

			Log.i(clsName, "jitter calculation");
			jitterCalculation();
			if (halted) {
				return null;
			}
			retrieveJitterResult();
			publishProgress(50);

			Log.i(clsName, "bandwidth calculation");
			boolean finish = bandwidthCalculation();
			publishProgress(100);

			// a task foi cancelada ou foi parado por um timer
			if (halted || !finish) {
				return null;
			}

			Log.d(clsName, "ProfileFull Finished");

			return network;
		} catch (InterruptedException e) {
			Log.w(clsName, e);
		} catch (IOException e) {
			Log.e(clsName, e.getMessage(), e);
		} catch (MissedEventException e) {
			Log.e(clsName, e.getMessage(), e);
		}

		publishProgress(100);
		return null;
	}

	/**
	 * Definição: RFC 4689 - defines jitter as “the absolute value of the difference between the Forwarding Delay of two consecutive received packets
	 * belonging to the same stream”. The jitter is important in real-time communications when the variation between delays can cause a negative
	 * impact to the server quality, such voice over IP services. Referencia: http://tools.ietf.org/html/rfc4689#section-3.2.5 Em resumo o jitter
	 * calcula os intervalos tempos entre o intervalo de tempo (corrente) e intervalo de tempo (anterior) e deve ser enviado num fluxo de taxa
	 * constante. #formula no servidor Intervalo de tempo (It) = Tempo_chegada - Tempo_anterior Jitter = it_atual - it_anterior (voce pode pegar a
	 * média, maximo e minimo) Sobre os resultados: Um jitter de 15ms é regular, abaixo de 5ms é excelente e acima de 15ms é ruim para o padrão VoIP.
	 * Seja esse site: http://www.onsip.com/tools/voip-test
	 * 
	 * @throws MissedEventException
	 * @throws IOException
	 * @throws InterruptedException
	 */
	private void jitterCalculation() throws IOException, MissedEventException, InterruptedException {
		ClientAbstract client = FactoryClient.getInstance(Protocol.UDP_EVENT);
		client.setReceiveDataEvent(new ReceiveDataEvent() {
			@Override
			public void receive(byte[] data, int offset, int read) {
				Log.d(clsName, "Jitter Finish");
			}
		});

		client.connect(server.getIp(), server.getJitterTestPort());

		for (int i = 0; i < 21; i++) {
			client.send(("jitter").getBytes());

			// bota 250ms para retorno
			// por causa do UDP que não tem controle de fluxo
			Thread.sleep(250);
		}

		client.close();
	}

	private void retrieveJitterResult() throws IOException, MissedEventException, InterruptedException {
		Thread.sleep(1500);
		final Semaphore mutex = new Semaphore(0);

		ClientAbstract client = FactoryClient.getInstance(Protocol.TCP_EVENT);
		client.setReceiveDataEvent(new ReceiveDataEvent() {
			@Override
			public void receive(byte[] data, int offset, int read) {
				Log.d(clsName, "Retrieve data from server for Jitter calcule");

				network.setJitter(Integer.parseInt(new String(data, offset, read)));
				// System.out.println(results.getJitter());

				mutex.release();
			}
		});

		client.connect(server.getIp(), server.getJitterRetrieveResultPort());
		client.send("get".getBytes());

		mutex.acquire();
		client.close();
	}

	private boolean bandwidthCalculation() throws IOException, MissedEventException, InterruptedException {
		final Semaphore mutex = new Semaphore(0);
		
		//begin download
		publishProgress(55);
		
		ClientAbstract client = FactoryClient.getInstance(Protocol.TCP_EVENT);
		client.setReceiveDataEvent(new ReceiveDataEvent() {
			private long countBytes = 0L;

			private byte endDown[] = "end_down".getBytes();
			private byte endSession[] = "end_session".getBytes();

			@Override
			public void receive(byte[] data, int offset, int read) {
				countBytes += (long) read;

				if (Util.containsArrays(data, endDown)) {
					// System.out.println("Bytes: "+countBytes);
					// bytes * 8bits / 7s * 1E+6 = X Mbits
					double bandwidth = ((double) (countBytes * 8L) / (double) (7.0 * 1E+6));
					network.setBandwidthDownload(String.valueOf(bandwidth));
					countBytes = 0L;
					mutex.release();
				} else if (Util.containsArrays(data, endSession)) {
					bandwidthDone = true;
					String dataBlock = new String(data, offset, read);
					String res[] = dataBlock.split(":");
					network.setBandwidthUpload(res[1]);

					mutex.release();
				}
			}
		});

		// timer for finish!
		Timer timeout = new Timer("Timeout Bandwidth");
		timeout.schedule(new TimerTask() {
			@Override
			public void run() {
				if (!bandwidthDone) {
					// para garantir que não vai travar nenhum semaphoro!
					mutex.release();
					mutex.release();
					Log.i(clsName, "Bandwidth Timeout...");
				}
			}
		}, 120000);// 120s de timeout

		Log.i(clsName, "bandwidth (download)");
		client.connect(server.getIp(), server.getBandwidthPort());
		client.send("down".getBytes());
		
		//wait finish the down...
		mutex.acquire();
		
		//begin upload
		publishProgress(75);

		if (halted) {
			timeout.cancel();
			return false;
		}

		Log.i(clsName, "bandwidth (upload)");
		client.send("up".getBytes());

		// faz upload! - 11s
		long timeExit = System.currentTimeMillis() + 11000;
		while (System.currentTimeMillis() < timeExit) {
			client.send(data);
		}
		client.send("end_up".getBytes());

		Log.i(clsName, "bandwidth (ended upload)");
		mutex.acquire();
		client.close();

		// cancela o timer
		timeout.cancel();

		return bandwidthDone;
	}
}